### PR TITLE
Change securesystemslib.dsse.Envelope.signatures to dict upstream

### PR DESCRIPTION
**Pull request by NicholasTanz**
Fixes: https://github.com/theupdateframework/python-tuf/issues/2564 

### Please verify and check that the pull request fulfils the following requirements:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

### CODE REVIEW

File: securesystemslib/dsse.py [hunk 1]
```diff
@@ -26,7 +26,7 @@ def __init__(
     ):
```
Comment (NicholasTanz) [inline securesystemslib/dsse.py:26]: I wasn't sure if I should change the last param from signatures: List[Signature] to signatures: Dict[str, Signature], so any information is appreciated

**Comment by NicholasTanz** [comment]
Hey @lukpueh , I was wondering if you could take a look at this pr when you get a chance? thanks

File: securesystemslib/dsse.py (hunk 1 above)
Comment (lukpueh) [reply securesystemslib/dsse.py:26]: Good question! Please do change the constructor argument to `Dict[str, Signature]` and implement the translation in `from_dict` and `to_dict`. This is also what we do with [python-tuf  `Metadata`](https://github.com/theupdateframework/python-tuf/blob/e1b52e7fd828c11159a12aef4259133c3f05accf/tuf/api/metadata.py#L125).

Comment (lukpueh) [reply securesystemslib/dsse.py:26]: I also just saw that we raise, if there are [duplicate (by keyid)](https://github.com/theupdateframework/python-tuf/blob/e1b52e7fd828c11159a12aef4259133c3f05accf/tuf/api/metadata.py#L189-L195) signatures there. I suggest to do the same in `Envelope.from_dict`.

**Review by lukpueh — CHANGES_REQUESTED** [review]
Thanks for the clean patch! I replied to your question inline. Please address and we can merge this. FYI I already tried out what changes this will need in tuf and in-toto, and it looks reasonable (see referenced commits above).

File: securesystemslib/dsse.py [hunk 2]
```diff
@@ -58,18 +61,23 @@ def from_dict(cls, data: dict) -> "Envelope":
...
-        signatures = []
+        signatures = {}
         for signature in data["signatures"]:
             signature["sig"] = b64dec(signature["sig"]).hex()
-            signatures.append(Signature.from_dict(signature))
+            signature = Signature.from_dict(signature)
+            if signature.keyid in signatures:
+                raise ValueError(
+                    f"Multiple signatures found for keyid {signature.keyid}"
+                )
```
Comment (lukpueh) [inline securesystemslib/dsse.py:71]: Thanks for the check. Would you mind adding a small test?

File: securesystemslib/dsse.py [hunk 3]
```diff
@@ -58,18 +61,23 @@ def from_dict(cls, data: dict) -> "Envelope":
...
+            signatures[signature.keyid] = signature
 
         return cls(payload, payload_type, signatures)
 
     def to_dict(self) -> dict:
         """Returns the JSON-serializable dictionary representation of self."""
 
         signatures = []
-        for signature in self.signatures:
+        for signature in list(self.signatures.values()):
```
Comment (lukpueh) [inline securesystemslib/dsse.py:80]: Is there a reason to convert to `list`? You could iterate directly over  `dict_values`. Same comment applies below.

Comment (NicholasTanz) [reply securesystemslib/dsse.py:80]: ah no - I just followed the format here [comment](https://github.com/theupdateframework/python-tuf/pull/2436#discussion_r1482773245) . Will change to just iterating over dict_values

Comment (lukpueh) [reply securesystemslib/dsse.py:80]: Yeah, that PR has a few such conversions for type checking reasons.

**Review by lukpueh — APPROVED** [review]
LGTM!

**Pull request merged**